pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp

```diff
@@ -38,27 +45,107 @@ typedef control_msgs::FollowJointTrajectoryGoal GoalType;
+static const std::string JOINTS_PARAMETER {"joints"};
+static const std::vector<std::string> JOINT_NAMES {"shoulder_to_right_arm", "shoulder_to_left_arm"};
+
+static constexpr double DEFAULT_GOAL_DURATION_SEC {1.0};
+static constexpr double DEFAULT_UPDATE_PERIOD_SEC {0.008};
+static constexpr double STOP_TRAJECTORY_DURATION_SEC {0.2};
+static constexpr double GOAL_TIME_TOLERANCE_SEC {0.01};
+static constexpr double TIME_SIMULATION_START_SEC {0.1};
+static constexpr double VELOCITY_COMPARISON_EPS {0.01};
+static constexpr double WAIT_FOR_ACTION_RESULT_TIMEOUT_SEC {5.0};
+
+static constexpr unsigned int SLEEP_TIME_MSEC {5};
+static constexpr unsigned int WAIT_FOR_ACTION_SERVER_TIMEOUT_MSEC {5000};
+static constexpr unsigned int WAIT_FOR_HOLD_FUTURE_MSEC {1000};
+static constexpr unsigned int WAIT_FOR_MOVEMENT_STARTED_MSEC {1000};
 
 static void progressInTime(const ros::Duration& period)
 {
-  ros::Time current_time{ros::Time::now()};
+  ros::Time current_time {ros::Time::now()};
   ros::Time::setNow(current_time + period);
 }
 
+static ros::Duration getGoalDuration(const GoalType &goal)
+{
+  return goal.trajectory.points.back().time_from_start;
+}
+
+static bool waitForUsingSystemTime(const std::function<bool()>& is_condition_fulfilled,
+                                   const std::chrono::milliseconds& timeout)
+{
+  const std::chrono::system_clock::time_point start {std::chrono::system_clock::now()};
+  do
+  {
+    if (is_condition_fulfilled())
+    {
+      return true;
+    }
+    if (std::chrono::system_clock::now() - start > timeout)
+    {
+      return false;
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(SLEEP_TIME_MSEC));
+  }
+  while (ros::ok());
+  return false;
+}
+
+template<typename T>
+bool isFutureReady(const std::future<T>& this_future)
+{
+  return this_future.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
+}
+
+/**
+ * @brief For simplicity this robot only updates one joint of the two joints defined in the urdf.
+ */
+class RobotMock
+{
+public:
+  RobotMock(HWInterface* hardware);
+
+  void update(const ros::Duration& period);
+
+  bool isMoving();
+
+private:
+  double pos_ = 0.0, vel_ = 0.0, eff_ = 0.0, cmd_ = 0.0;
+  double pos2_ = 0.0, vel2_ = 0.0, eff2_ = 0.0, cmd2_ = 0.0;
+};
+
+RobotMock::RobotMock(HWInterface* hardware)
+{
+  hardware_interface::JointStateHandle jsh {JOINT_NAMES.at(0), &pos_, &vel_, &eff_};
+  hardware_interface::JointHandle jh {jsh, &cmd_};
+  hardware->registerHandle(jh);
+  hardware_interface::JointStateHandle jsh2 {JOINT_NAMES.at(1), &pos2_, &vel2_, &eff2_};
+  hardware_interface::JointHandle jh2 {jsh2, &cmd2_};
+  hardware->registerHandle(jh2);
+}
+
+void RobotMock::update(const ros::Duration& period)
+{
+vel_ = (cmd_ - pos_) / period.toSec();
+pos_ = cmd_;
+}
+
+bool RobotMock::isMoving()
+{
+  return std::abs(vel_) > VELOCITY_COMPARISON_EPS;
+}
+
 /**
  * @brief Test fixture class for the unit-test of the PilzJointTrajectoryController.
```

Comment:
The `PilzJointTrajectoryControllerTest` class consists of quite a number of member variables and helper functions, partially using the member variables. Also, the `PilzJointTrajectoryControllerIsExecutingTest` is derived from `PilzJointTrajectoryControllerTest` and uses the members and the helper functions too.
Furthermore, the test suite consists of a "wall of code" which makes it difficult to understand and read.
Therefore, I suggest the following:
- Separate the members and helper functions used by both test classes (`PilzJointTrajectoryControllerTest` and `PilzJointTrajectoryControllerIsExecutingTest`) into a separate independent class.
- Put the new class into a separate header. 
- Then include the new helper-class as member in both test classes.
- Then only derive `PilzJointTrajectoryControllerIsExecutingTest` from `testing::WithParamInterface<InvokeIsExecuting>` and nothing else.

**Please note:**
- Inheritance is not the only way to re-use helper functions across different classes, etc.

Reply:
I would not swap out all functions & members into a single helper class, but make it dependent on the context.
I just started implementing a `PJTCManagerMock` (see comment above). After that I will try to extract further components in order to reduced the amount of code in the test suite.

Reply:
> I would not swap out all functions & members into a single helper class, but make it dependent on the context.

No, only separate the functions & member into a helper class which actually create a logical unit. I also want to emphasize that I thjnk that the inheritance model used here is faulty from a clean code perspective.

> After that I will try to extract further components in order to reduced the amount of code in the test suite.

OK, I will wait ;)

Reply:
Done. You can recheck it if you want.